# Understanding Covid19 Case Counts using machine learning

In [72]:
# used library imports
import pandas as pd

### Covid Data from Our World in Data

Last downloaded: 16th of May 2024

https://github.com/owid/covid-19-data/blob/master/public/data/owid-covid-data.csv

In [73]:
df = pd.read_csv('resources/owid-covid-data.csv')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 397677 entries, 0 to 397676
Data columns (total 67 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   iso_code                                    397677 non-null  object 
 1   continent                                   378618 non-null  object 
 2   location                                    397677 non-null  object 
 3   date                                        397677 non-null  object 
 4   total_cases                                 358581 non-null  float64
 5   new_cases                                   386444 non-null  float64
 6   new_cases_smoothed                          385214 non-null  float64
 7   total_deaths                                336105 non-null  float64
 8   new_deaths                                  386794 non-null  float64
 9   new_deaths_smoothed                         385564 non-null  float64
 

### Inspecting the Data

Findings from a first look at the data:
- need to decide wether to use smoothed variables or not, can't use both
- total tests probably redundant
- total vaccinations are redundant if we have people vaccinated, fully vaccinated and boosters
- some columns have high numbers of missing values, like handwashing_facilities and smokers
- for icu and hospital patients, per million might be better, due to comparability



In [74]:
df

iso_code continent     location        date  total_cases  new_cases  \
0           AFG      Asia  Afghanistan  2020-01-05          NaN        0.0   
1           AFG      Asia  Afghanistan  2020-01-06          NaN        0.0   
2           AFG      Asia  Afghanistan  2020-01-07          NaN        0.0   
3           AFG      Asia  Afghanistan  2020-01-08          NaN        0.0   
4           AFG      Asia  Afghanistan  2020-01-09          NaN        0.0   
...         ...       ...          ...         ...          ...        ...   
397672      ZWE    Africa     Zimbabwe  2024-04-24     266359.0        0.0   
397673      ZWE    Africa     Zimbabwe  2024-04-25     266359.0        0.0   
397674      ZWE    Africa     Zimbabwe  2024-04-26     266359.0        0.0   
397675      ZWE    Africa     Zimbabwe  2024-04-27     266359.0        0.0   
397676      ZWE    Africa     Zimbabwe  2024-04-28     266359.0        0.0   

        new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  \
0                      NaN           NaN         0.0                  NaN   
1                      NaN           NaN         0.0                  NaN   
2                      NaN           NaN         0.0                  NaN   
3                      NaN           NaN         0.0                  NaN   
4                      NaN           NaN         0.0                  NaN   
...                    ...           ...         ...                  ...   
397672                 0.0        5740.0         0.0                  0.0   
397673                 0.0        5740.0         0.0                  0.0   
397674                 0.0        5740.0         0.0                  0.0   
397675                 0.0        5740.0         0.0                  0.0   
397676                 0.0        5740.0         0.0                  0.0   

        ...  male_smokers  handwashing_facilities  hospital_beds_per_thousand  \
0       ...           NaN                  37.746                         0.5   
1       ...           NaN                  37.746                         0.5   
2       ...           NaN                  37.746                         0.5   
3       ...           NaN                  37.746                         0.5   
4       ...           NaN                  37.746                         0.5   
...     ...           ...                     ...                         ...   
397672  ...          30.7                  36.791                         1.7   
397673  ...          30.7                  36.791                         1.7   
397674  ...          30.7                  36.791                         1.7   
397675  ...          30.7                  36.791                         1.7   
397676  ...          30.7                  36.791                         1.7   

        life_expectancy  human_development_index  population  \
0                 64.83                    0.511  41128772.0   
1                 64.83                    0.511  41128772.0   
2                 64.83                    0.511  41128772.0   
3                 64.83                    0.511  41128772.0   
4                 64.83                    0.511  41128772.0   
...                 ...                      ...         ...   
397672            61.49                    0.571  16320539.0   
397673            61.49                    0.571  16320539.0   
397674            61.49                    0.571  16320539.0   
397675            61.49                    0.571  16320539.0   
397676            61.49                    0.571  16320539.0   

        excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                        NaN                          NaN   
1                                        NaN                          NaN   
2                                        NaN                          NaN   
3                                        NaN                          NaN   
4                 

In [75]:
df.describe()

total_cases     new_cases  new_cases_smoothed  total_deaths  \
count  3.585810e+05  3.864440e+05        3.852140e+05  3.361050e+05   
mean   7.560483e+06  8.503695e+03        8.530716e+03  9.096952e+04   
std    4.476474e+07  2.365363e+05        8.919575e+04  4.608855e+05   
min    1.000000e+00  0.000000e+00        0.000000e+00  1.000000e+00   
25%    9.631000e+03  0.000000e+00        0.000000e+00  1.380000e+02   
50%    8.158100e+04  0.000000e+00        1.557100e+01  1.407000e+03   
75%    9.021610e+05  0.000000e+00        3.748570e+02  1.275300e+04   
max    7.753798e+08  4.423623e+07        6.319461e+06  7.047396e+06   

          new_deaths  new_deaths_smoothed  total_cases_per_million  \
count  386794.000000        385564.000000            358581.000000   
mean       76.169995            76.409550            115014.660831   
std      1405.682671           527.419277            162125.095103   
min         0.000000             0.000000                 0.000000   
25%         0.000000             0.000000              3040.311000   
50%         0.000000             0.000000             33973.672000   
75%         0.000000             3.714000            157521.073000   
max    103719.000000         14817.000000            770966.075000   

       new_cases_per_million  new_cases_smoothed_per_million  \
count          386444.000000                   385214.000000   
mean              129.532618                      129.933273   
std              1526.641288                      565.267399   
min                 0.000000                        0.000000   
25%                 0.000000                        0.000000   
50%                 0.000000                        3.801000   
75%                 0.000000                       64.734000   
max            240325.866000                    34332.267000   

       total_deaths_per_million  ...   male_smokers  handwashing_facilities  \
count             336105.000000  ...  228740.000000           151351.000000   
mean                 929.946747  ...      32.910900               50.790431   
std                 1139.151038  ...      13.572257               31.952834   
min                    0.000000  ...       7.700000                1.188000   
25%                   66.820000  ...      22.600000               20.859000   
50%                  433.799000  ...      33.100000               49.839000   
75%                 1493.483000  ...      41.300000               82.502000   
max                 6485.570000  ...      78.100000              100.000000   

       hospital_beds_per_thousand  life_expectancy  human_development_index  \
count               272868.000000    366406.000000            299646.000000   
mean                     3.097428        73.710804                 0.722574   
std                      2.548117         7.394839                 0.149003   
min                      0.100000        53.280000                 0.394000   
25%                      1.300000        69.590000                 0.602000   
50%                      2.500000        75.050000                 0.740000   
75%                      4.200000        79.460000                 0.829000   
max                     13.800000        86.750000                 0.957000   

         population  excess_mortality_cumulative_absolute  \
count  3.976770e+05                          1.327600e+04   
mean   1.293155e+08                          5.565216e+04   
std    6.635272e+08                          1.554990e+05   
min    4.700000e+01                         -3.772610e+04   
25%    4.490020e+05                          1.776251e+02   
50%    5.882259e+06                          6.806349e+03   
75%    2.830170e+07                          3.894165e+04   
max    7.975105e+09                          1.345330e+06   

       excess_mortality_cumulative  excess_mortality  \
count                 13276.000000      13276.000000   
mean                      9.762172         10.986410   
std               

Selecting a subset of columns that can be used for predictions

In [76]:
df_subset = df[['continent','location','date','new_cases_per_million','new_deaths_per_million','reproduction_rate','icu_patients_per_million','hosp_patients_per_million','new_tests_per_thousand','people_vaccinated_per_hundred','people_fully_vaccinated_per_hundred','total_boosters_per_hundred','stringency_index','population_density','median_age','aged_65_older','aged_70_older','gdp_per_capita','cardiovasc_death_rate','diabetes_prevalence','female_smokers','male_smokers','hospital_beds_per_thousand','life_expectancy','human_development_index',]]
df_subset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 397677 entries, 0 to 397676
Data columns (total 25 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   continent                            378618 non-null  object 
 1   location                             397677 non-null  object 
 2   date                                 397677 non-null  object 
 3   new_cases_per_million                386444 non-null  float64
 4   new_deaths_per_million               386794 non-null  float64
 5   reproduction_rate                    184817 non-null  float64
 6   icu_patients_per_million             38837 non-null   float64
 7   hosp_patients_per_million            40377 non-null   float64
 8   new_tests_per_thousand               75403 non-null   float64
 9   people_vaccinated_per_hundred        79959 non-null   float64
 10  people_fully_vaccinated_per_hundred  76865 non-null   float64
 11  total_booster

In [77]:
df_subset.isnull().sum()

continent                               19059
location                                    0
date                                        0
new_cases_per_million                   11233
new_deaths_per_million                  10883
reproduction_rate                      212860
icu_patients_per_million               358840
hosp_patients_per_million              357300
new_tests_per_thousand                 322274
people_vaccinated_per_hundred          317718
people_fully_vaccinated_per_hundred    320812
total_boosters_per_hundred             345300
stringency_index                       200385
population_density                      59412
median_age                              83279
aged_65_older                           94084
aged_70_older                           86431
gdp_per_capita                          89356
cardiovasc_death_rate                   88789
diabetes_prevalence                     72815
female_smokers                         165785
male_smokers                      

In [78]:
df_subset.describe()

new_cases_per_million  new_deaths_per_million  reproduction_rate  \
count          386444.000000           386794.000000      184817.000000   
mean              129.532618                0.806203           0.911495   
std              1526.641288                7.211076           0.399925   
min                 0.000000                0.000000          -0.070000   
25%                 0.000000                0.000000           0.720000   
50%                 0.000000                0.000000           0.950000   
75%                 0.000000                0.000000           1.140000   
max            240325.866000              906.413000           5.870000   

       icu_patients_per_million  hosp_patients_per_million  \
count              38837.000000               40377.000000   
mean                  15.767063                 126.813391   
std                   22.829584                 151.347732   
min                    0.000000                   0.000000   
25%                    2.378000                  32.016000   
50%                    6.527000                  74.968000   
75%                   18.973000                 160.568000   
max                  180.675000                1526.846000   

       new_tests_per_thousand  people_vaccinated_per_hundred  \
count            75403.000000                   79959.000000   
mean                 3.272466                      53.204774   
std                  9.033843                      29.480966   
min                  0.000000                       0.000000   
25%                  0.286000                      27.260000   
50%                  0.971000                      63.990000   
75%                  2.914000                      77.745000   
max                531.062000                     129.070000   

       people_fully_vaccinated_per_hundred  total_boosters_per_hundred  \
count                         76865.000000                52377.000000   
mean                             48.368835                   36.023039   
std                              29.139236                   30.397027   
min                               0.000000                    0.000000   
25%                              20.860000                    5.760000   
50%                              57.080000                   35.310000   
75%                              73.600000                   57.700000   
max                             126.890000                  150.470000   

       stringency_index  ...  aged_65_older  aged_70_older  gdp_per_capita  \
count     197292.000000  ...  303593.000000  311246.000000   308321.000000   
mean          42.791492  ...       8.703831       5.500781    18961.090464   
std           24.867260  ...       6.097274       4.138575    19875.492581   
min            0.000000  ...       1.144000       0.526000      661.240000   
25%           22.220000  ...       3.526000       2.085000     3823.194000   
50%           42.590000  ...       6.378000       3.871000    12294.876000   
75%           62.040000  ...      13.928000       8.643000    27216.445000   
max          100.000000  ...      27.049000      18.493000   116935.600000   

       cardiovasc_death_rate  diabetes_prevalence  female_smokers  \
count          308888.000000        324862.000000   231892.000000   
mean              264.360275             8.560196       10.795876   
std               120.886118             4.940448       10.781817   
min                79.370000             0.990000        0.100000   
25%               175.695000             5.350000        1.900000   
50%               245.465000             7.200000        6.300000   
75%               333.436000            10.790000       19.300000   
max               724.417000            30.530000       44.000000   

        male_smokers  hospital_beds_per_thousand  life_expectancy  \
count  228740.000000               272868.000000    366406.000000   
mean       32.910900                    3.097428        73.710804

overall there are a lot of null values especially for vaccinations, so we need to check if that's because some countries just don't report vaccinations, or if those values are all from a time where there were no vaccinations, also if there are gaps in the reports, so fill them with 0 till the first vaccination and then last observation carried forward

In [79]:
df_subset['continent'].unique()

array(['Asia', nan, 'Europe', 'Africa', 'Oceania', 'North America',
       'South America'], dtype=object)

In [80]:
df_cleaned = df_subset.dropna(subset=['continent'])
df_cleaned.loc[:, 'new_cases_per_million']= df_cleaned['new_cases_per_million'].fillna(0)
df_cleaned.loc[:, 'new_deaths_per_million']= df_cleaned['new_deaths_per_million'].fillna(0)
df_cleaned.loc[:, 'reproduction_rate']= df_cleaned['reproduction_rate'].fillna(0)
df_cleaned.loc[:, 'icu_patients_per_million']= df_cleaned['icu_patients_per_million'].fillna(0)
df_cleaned.loc[:, 'hosp_patients_per_million']= df_cleaned['hosp_patients_per_million'].fillna(0)
df_cleaned.loc[:, 'new_tests_per_thousand']= df_cleaned['new_tests_per_thousand'].fillna(0)
df_cleaned.loc[:, 'stringency_index']= df_cleaned['stringency_index'].fillna(0)

df_cleaned

continent     location        date  new_cases_per_million  \
0           Asia  Afghanistan  2020-01-05                    0.0   
1           Asia  Afghanistan  2020-01-06                    0.0   
2           Asia  Afghanistan  2020-01-07                    0.0   
3           Asia  Afghanistan  2020-01-08                    0.0   
4           Asia  Afghanistan  2020-01-09                    0.0   
...          ...          ...         ...                    ...   
397672    Africa     Zimbabwe  2024-04-24                    0.0   
397673    Africa     Zimbabwe  2024-04-25                    0.0   
397674    Africa     Zimbabwe  2024-04-26                    0.0   
397675    Africa     Zimbabwe  2024-04-27                    0.0   
397676    Africa     Zimbabwe  2024-04-28                    0.0   

        new_deaths_per_million  reproduction_rate  icu_patients_per_million  \
0                          0.0                0.0                       0.0   
1                          0.0                0.0                       0.0   
2                          0.0                0.0                       0.0   
3                          0.0                0.0                       0.0   
4                          0.0                0.0                       0.0   
...                        ...                ...                       ...   
397672                     0.0                0.0                       0.0   
397673                     0.0                0.0                       0.0   
397674                     0.0                0.0                       0.0   
397675                     0.0                0.0                       0.0   
397676                     0.0                0.0                       0.0   

        hosp_patients_per_million  new_tests_per_thousand  \
0                             0.0                     0.0   
1                             0.0                     0.0   
2                             0.0                     0.0   
3                             0.0                     0.0   
4                             0.0                     0.0   
...                           ...                     ...   
397672                        0.0                     0.0   
397673                        0.0                     0.0   
397674                        0.0                     0.0   
397675                        0.0                     0.0   
397676                        0.0                     0.0   

        people_vaccinated_per_hundred  ...  aged_65_older  aged_70_older  \
0                                 NaN  ...          2.581          1.337   
1                                 NaN  ...          2.581          1.337   
2                                 NaN  ...          2.581          1.337   
3                                 NaN  ...          2.581          1.337   
4                                 NaN  ...          2.581          1.337   
...                               ...  ...            ...            ...   
397672                            NaN  ...          2.822          1.882   
397673                            NaN  ...          2.822          1.882   
397674                            NaN  ...          2.822          1.882   
397675                            NaN  ...          2.822          1.882   
397676                            NaN  ...          2.822          1.882   

        gdp_per_capita  cardiovasc_death_rate  diabetes_prevalence  \
0             1803.987                597.029                 9.59   
1             1803.987                597.029                 9.59   
2             1803.987                597.029                 9.59   
3             1803.987                597.029                 9.59   
4             1803.987                597.029                 9.59   
...                ...                    ...                  ...   
397672        1899.775                307.846                 1.82   
397673        1899.775             

In [81]:
#sort by 'location' and 'date'
df_cleaned = df_cleaned.sort_values(by=['location', 'date'])

# Use groupby on 'location' and apply forward fill
df_cleaned['people_vaccinated_per_hundred'] = df_cleaned.groupby('location')['people_vaccinated_per_hundred'].ffill()
df_cleaned['people_fully_vaccinated_per_hundred'] = df_cleaned.groupby('location')['people_fully_vaccinated_per_hundred'].ffill()
df_cleaned['total_boosters_per_hundred'] = df_cleaned.groupby('location')['total_boosters_per_hundred'].ffill()

#fill the rest of values with 0 (those are values before the first value, therefore they are 0)
df_cleaned.loc[:, 'people_vaccinated_per_hundred']= df_cleaned['people_vaccinated_per_hundred'].fillna(0)
df_cleaned.loc[:, 'people_fully_vaccinated_per_hundred']= df_cleaned['people_fully_vaccinated_per_hundred'].fillna(0)
df_cleaned.loc[:, 'total_boosters_per_hundred']= df_cleaned['total_boosters_per_hundred'].fillna(0)

df_cleaned

continent     location        date  new_cases_per_million  \
0           Asia  Afghanistan  2020-01-05                    0.0   
1           Asia  Afghanistan  2020-01-06                    0.0   
2           Asia  Afghanistan  2020-01-07                    0.0   
3           Asia  Afghanistan  2020-01-08                    0.0   
4           Asia  Afghanistan  2020-01-09                    0.0   
...          ...          ...         ...                    ...   
397672    Africa     Zimbabwe  2024-04-24                    0.0   
397673    Africa     Zimbabwe  2024-04-25                    0.0   
397674    Africa     Zimbabwe  2024-04-26                    0.0   
397675    Africa     Zimbabwe  2024-04-27                    0.0   
397676    Africa     Zimbabwe  2024-04-28                    0.0   

        new_deaths_per_million  reproduction_rate  icu_patients_per_million  \
0                          0.0                0.0                       0.0   
1                          0.0                0.0                       0.0   
2                          0.0                0.0                       0.0   
3                          0.0                0.0                       0.0   
4                          0.0                0.0                       0.0   
...                        ...                ...                       ...   
397672                     0.0                0.0                       0.0   
397673                     0.0                0.0                       0.0   
397674                     0.0                0.0                       0.0   
397675                     0.0                0.0                       0.0   
397676                     0.0                0.0                       0.0   

        hosp_patients_per_million  new_tests_per_thousand  \
0                             0.0                     0.0   
1                             0.0                     0.0   
2                             0.0                     0.0   
3                             0.0                     0.0   
4                             0.0                     0.0   
...                           ...                     ...   
397672                        0.0                     0.0   
397673                        0.0                     0.0   
397674                        0.0                     0.0   
397675                        0.0                     0.0   
397676                        0.0                     0.0   

        people_vaccinated_per_hundred  ...  aged_65_older  aged_70_older  \
0                                0.00  ...          2.581          1.337   
1                                0.00  ...          2.581          1.337   
2                                0.00  ...          2.581          1.337   
3                                0.00  ...          2.581          1.337   
4                                0.00  ...          2.581          1.337   
...                               ...  ...            ...            ...   
397672                          39.45  ...          2.822          1.882   
397673                          39.45  ...          2.822          1.882   
397674                          39.45  ...          2.822          1.882   
397675                          39.45  ...          2.822          1.882   
397676                          39.45  ...          2.822          1.882   

        gdp_per_capita  cardiovasc_death_rate  diabetes_prevalence  \
0             1803.987                597.029                 9.59   
1             1803.987                597.029                 9.59   
2             1803.987                597.029                 9.59   
3             1803.987                597.029                 9.59   
4             1803.987                597.029                 9.59   
...                ...                    ...                  ...   
397672        1899.775                307.846                 1.82   
397673        1899.775             

In [82]:
#df_cleaned[df_cleaned['location'] == 'Austria']
#the locations with missing values for population_density are small islands, so we can drop them
print(df_cleaned[df_cleaned['population_density'].isna()]['location'].unique())
df_cleaned.dropna(subset=['population_density'],inplace=True)


['Anguilla' 'Bonaire Sint Eustatius and Saba' 'Cook Islands' 'England'
 'Falkland Islands' 'French Guiana' 'Guadeloupe' 'Guernsey' 'Jersey'
 'Martinique' 'Mayotte' 'Montserrat' 'Niue' 'Northern Cyprus'
 'Northern Ireland' 'Pitcairn' 'Reunion' 'Saint Barthelemy' 'Saint Helena'
 'Saint Pierre and Miquelon' 'Scotland' 'South Sudan' 'Syria' 'Taiwan'
 'Tokelau' 'Vatican' 'Wales' 'Wallis and Futuna' 'Western Sahara']


In [83]:
#same for median_age the locations, that have missing values are small islands, so we can drop them
print(df_cleaned[df_cleaned['median_age'].isna()]['location'].unique())
df_cleaned.dropna(subset=['median_age'],inplace=True)

['American Samoa' 'Andorra' 'Bermuda' 'British Virgin Islands'
 'Cayman Islands' 'Dominica' 'Faeroe Islands' 'Gibraltar' 'Greenland'
 'Isle of Man' 'Kosovo' 'Liechtenstein' 'Marshall Islands' 'Monaco'
 'Nauru' 'Northern Mariana Islands' 'Palau' 'Saint Kitts and Nevis'
 'Saint Martin (French part)' 'San Marino' 'Sint Maarten (Dutch part)'
 'Turks and Caicos Islands' 'Tuvalu']


In [84]:
#locations with missing gdp per capita are countries with limited information like, North Korea, Cuba or Somalia, so we drop them
print(df_cleaned[df_cleaned['gdp_per_capita'].isna()]['location'].unique())
df_cleaned.dropna(subset=['gdp_per_capita'],inplace=True)

['Cuba' 'Curacao' 'French Polynesia' 'Guam' 'New Caledonia' 'North Korea'
 'Somalia' 'United States Virgin Islands']


In [85]:
#only puerto rico has no human development index, so we drop it
print(df_cleaned[df_cleaned['human_development_index'].isna()]['location'].unique())
df_cleaned.dropna(subset=['human_development_index'],inplace=True)

['Aruba' 'Macao' 'Puerto Rico']


In [86]:
#missing values for 70 or older are assumed to be null
df_cleaned['aged_70_older'].fillna(0, inplace=True)
#same for cardiovasc_death_rate
df_cleaned.dropna(subset=['cardiovasc_death_rate'],inplace=True)
#smokers are filled with the median value (they are also mainly small 3rd world countries, so maybe should be dropped)
# Calculate the median of 'female_smokers' and 'male_smokers'
female_smokers_median = df_cleaned['female_smokers'].median()
male_smokers_median = df_cleaned['male_smokers'].median()

# Fill the missing values with the calculated median
df_cleaned['female_smokers'].fillna(female_smokers_median, inplace=True)
df_cleaned['male_smokers'].fillna(male_smokers_median, inplace=True)

#hospital beds are also small countries or poor countries, so we assume 0 
df_cleaned['hospital_beds_per_thousand'].fillna(0, inplace=True)
#only puerto rico has no human development index, so we drop it
df_cleaned.dropna(subset=['human_development_index'],inplace=True)

#df_cleaned.to_csv('resources/covid_data_cleaned.csv')
df_cleaned

continent     location        date  new_cases_per_million  \
0           Asia  Afghanistan  2020-01-05                    0.0   
1           Asia  Afghanistan  2020-01-06                    0.0   
2           Asia  Afghanistan  2020-01-07                    0.0   
3           Asia  Afghanistan  2020-01-08                    0.0   
4           Asia  Afghanistan  2020-01-09                    0.0   
...          ...          ...         ...                    ...   
397672    Africa     Zimbabwe  2024-04-24                    0.0   
397673    Africa     Zimbabwe  2024-04-25                    0.0   
397674    Africa     Zimbabwe  2024-04-26                    0.0   
397675    Africa     Zimbabwe  2024-04-27                    0.0   
397676    Africa     Zimbabwe  2024-04-28                    0.0   

        new_deaths_per_million  reproduction_rate  icu_patients_per_million  \
0                          0.0                0.0                       0.0   
1                          0.0                0.0                       0.0   
2                          0.0                0.0                       0.0   
3                          0.0                0.0                       0.0   
4                          0.0                0.0                       0.0   
...                        ...                ...                       ...   
397672                     0.0                0.0                       0.0   
397673                     0.0                0.0                       0.0   
397674                     0.0                0.0                       0.0   
397675                     0.0                0.0                       0.0   
397676                     0.0                0.0                       0.0   

        hosp_patients_per_million  new_tests_per_thousand  \
0                             0.0                     0.0   
1                             0.0                     0.0   
2                             0.0                     0.0   
3                             0.0                     0.0   
4                             0.0                     0.0   
...                           ...                     ...   
397672                        0.0                     0.0   
397673                        0.0                     0.0   
397674                        0.0                     0.0   
397675                        0.0                     0.0   
397676                        0.0                     0.0   

        people_vaccinated_per_hundred  ...  aged_65_older  aged_70_older  \
0                                0.00  ...          2.581          1.337   
1                                0.00  ...          2.581          1.337   
2                                0.00  ...          2.581          1.337   
3                                0.00  ...          2.581          1.337   
4                                0.00  ...          2.581          1.337   
...                               ...  ...            ...            ...   
397672                          39.45  ...          2.822          1.882   
397673                          39.45  ...          2.822          1.882   
397674                          39.45  ...          2.822          1.882   
397675                          39.45  ...          2.822          1.882   
397676                          39.45  ...          2.822          1.882   

        gdp_per_capita  cardiovasc_death_rate  diabetes_prevalence  \
0             1803.987                597.029                 9.59   
1             1803.987                597.029                 9.59   
2             1803.987                597.029                 9.59   
3             1803.987                597.029                 9.59   
4             1803.987                597.029                 9.59   
...                ...                    ...                  ...   
397672        1899.775                307.846                 1.82   
397673        1899.775             

In [87]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
Index: 282307 entries, 0 to 397676
Data columns (total 25 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   continent                            282307 non-null  object 
 1   location                             282307 non-null  object 
 2   date                                 282307 non-null  object 
 3   new_cases_per_million                282307 non-null  float64
 4   new_deaths_per_million               282307 non-null  float64
 5   reproduction_rate                    282307 non-null  float64
 6   icu_patients_per_million             282307 non-null  float64
 7   hosp_patients_per_million            282307 non-null  float64
 8   new_tests_per_thousand               282307 non-null  float64
 9   people_vaccinated_per_hundred        282307 non-null  float64
 10  people_fully_vaccinated_per_hundred  282307 non-null  float64
 11  total_boosters_per

convert date to datetime

In [88]:
df_cleaned['date'] = pd.to_datetime(df_cleaned['date'])

In [89]:
df_cleaned.to_csv('resources/covid_data_cleaned.csv', index=False)

confounder selection:
basesd on the pretreatment criterion, the following variables should be controlled for:
- population_density
- median_age
- gdp_per_capita
- life_expectancy
- human_development
- cardiovasc_death_rate
- diabetes_prevalence
- aged_65_older
- aged_70_older
- female_smokers
- male_smokers
based on the disjunctive cause criterion, the following variables should be controlled for:
maybe:
- new_deaths_per_million
proxies for unmeasured confounders:
- maybe use the season as a proxy for unmeasured confounders (like weather, holidays, etc.)

DoubleML to estimate the causal effect of the vaccination rate and stringency on the number of new cases per million

TODO for each continent 

This is an attempt at doing this doubleml manually

The value for theta for the custom implementation is 1.578 and theta for the doubleml implementation is 1.618, so the results are the quite similar, espacially when looking at the 2.5% and 97.5% quantiles, which are 1.517 and 1.720

trying out the econml package

Analysis with time shift